### PR TITLE
Add feature to enqueue dependents at the front of queues

### PR DESCRIPTION
**Pull request by jtfidje**
Suggested solution to #1690.

### CODE REVIEW

**Comment by selwin** [comment]
@jtfidje looks good. Do you mind updating the docs to add `enqueue_at_front` parameter to `Dependency` class? Also please make sure all the tests pass.

**Comment by jtfidje** [comment]
Aah yes - forgot docs! 

All tests passed when running `run_tests_in_docker.sh` - is there a way to run the CI tests locally as well?

**Comment by selwin** [comment]
@jtfidje can you check your tests and ensure that they pass? I'll merge if they can pass the tests.

File: docs/docs/index.md [hunk 1]
```diff
@@ -168,6 +168,8 @@ The `Dependency(jobs=...)` parameter accepts:
 - a Job object
 - an iteratable of job id strings and/or Job objects
```
Comment (selwin) [inline docs/docs/index.md:170]: Can you add this here (in the list of parameters)?
```
`enqueue_at_front` boolean parameter to put dependents at the front when they are enqueued
```

**Comment by jtfidje** [comment]
Looks like this PR can/will also benefit from #1693 - maybe we should wait and see?

**Comment by selwin** [comment]
I'd rather get this PR merged in first. We don't know when https://github.com/rq/rq/issues/1693 will be done.

If you can make all the tests pass, I'll merge this in and release a patch release along with a few bug fixes and improvements sometime in the next week or two.

**Comment by jtfidje** [comment]
No problem :) I'm facing a problem in my test tho: It seems that I cannot trust that the list of queued jobs I retrieve from Redis is ordered correctly. Even though I queue my `job_2` with `at_front=True`, the list I get back does not always have `job_2` at index 0. 
I've also tried to change my test so that I call `.work(...)` two times, one for parent and then once more to get the next job, but this has the same issue. Sometimes my `job_2` which is supposed to be at the front of the queue gets pulled and finished, but sometimes `job_1` gets pulled instead.

**Comment by selwin** [comment]
Ah ok, if that’s the problem, there’s a simple fix for that.

After you create the parent job along with the dependents, create another regular job so the queue is not empty.

Afterwards, run worker.work(max_jobs=1). This way, the one dependent will always be enqueued to the end of the queue and the other one always to the front.

This way we don’t depend on the order at which the dependents get enqueued.

**Comment by jtfidje** [comment]
Thanks for your time and quick responses!

I think that I don't quite follow what you mean, or maybe I've written a bad test. If we have the following setup:
```python
q.empty()
parent_job = q.enqueue(say_hello)
job_1 = q.enqueue(
    say_hello,
    job_id='fake_job_id_1',
    depends_on=Dependency(jobs=[parent_job])
)
job_2 = q.enqueue(
    say_hello,
    job_id='fake_job_id_2',
    depends_on=Dependency(jobs=[parent_job],enqueue_at_front=True)
)
w.work(burst=True, max_jobs=1)
```

I need to reliably test that `job_2` will run before `job_1`. I did this with the following assertion:

```python
self.assertEqual(q.job_ids, ["fake_job_id_2", "fake_job_id_1"])
```

How should I do this instead?

**Comment by selwin** [comment]
Try doing this:
```python
parent_job = q.enqueue(say_hello)
job_1 = q.enqueue(
    say_hello,
    job_id='fake_job_id_1',
    depends_on=Dependency(jobs=[parent_job])
)
job_2 = q.enqueue(
    say_hello,
    job_id='fake_job_id_2',
    depends_on=Dependency(jobs=[parent_job],enqueue_at_front=True)
)

q.enqueue(say_hello) # This is a filler job that will act as a separator for jobs, one will be enqueued at front while the other one at the end of the queue
w.work(burst=True, max_jobs=1)
```

**Comment by jtfidje** [comment]
I've added your suggestion and run the test three times. This is the result from `q.job_ids` in order:

```
['1bf9a09a-9b8c-4c31-9737-b61b970645bc', 'fake_job_id_1', 'fake_job_id_2']
['cd9bab30-9e45-4f03-b234-3c1ea080eed8', 'fake_job_id_2', 'fake_job_id_1']
['a2669abd-3eb7-42f5-b805-a122b88f11e9', 'fake_job_id_1', 'fake_job_id_2']
```

It seems the order is still random after `w.work(burst=True, max_jobs=1)` has been executed. I'm at a loss for how I could test if `at_front=True` is actually performed.

**Comment by selwin** [comment]
Then it seems like enqueue_at_front did not work. Did you check the logic and whether the enqueue_at_front variable was saved properly?

**Comment by selwin** [comment]
To recap, it seems like the issue is not in the random ordering but the job was never enqueued at front.

**Comment by jtfidje** [comment]
I'll dive back into the code again then :) Again, thank you for your time :)

**Comment by jtfidje** [comment]
You were correct - I had overlooked something critical. It now looks like everything is ok :) Please double check tho 😅

**Comment by selwin** [comment]
Sorry I missed your message. Will check in a bit.

**Comment by selwin** [comment]
`job.meta` is actually used for storing non core RQ data. Storing `enqueue_at_front` attribute on the actual job hash is actually the correct way.

If `job.enqueue_at_front` was not persisted correctly, you should look at:
1. `job.to_dict()` - serializes job attributes to be saved in Redis. Tackle this first so `enqueue_at_front` attribute gets correctly saved. https://github.com/rq/rq/blob/master/rq/job.py#L616
2. `job.restore()` - populates job attributes based on Redis hash. After it's saved properly, use `Job.fetch()` to test that `enqueue_at_front` gets properly restored. https://github.com/rq/rq/blob/master/rq/job.py#L545

**Comment by jtfidje** [comment]
Oh I understand! I'll fix it right away :)

File: rq/queue.py [hunk 2]
```diff
@@ -325,9 +325,6 @@ def create_job(self, func, args=None, kwargs=None, timeout=None,
             job.retries_left = retry.max
             job.retry_intervals = retry.intervals
 
-        if isinstance(depends_on, Dependency):
```
Comment (jtfidje) [inline rq/queue.py:328]: I took the liberty and removed these lines as this check is already done inside job.create.

**Comment by jtfidje** [comment]
Hi! I apologize for taking so long to fix this. Here is an updated proposal :)

I also took the liberty to remove some duplicate code - see: https://github.com/rq/rq/pull/1696/files#r954568688

**Comment by selwin** [comment]
Hey sorry I totally missed this update to your PR. This looks good, thanks!

**Comment by selwin** [comment]
Now I'm just re-running all the tests and if they pass, I can merge this in.

**Comment by jtfidje** [comment]
No problem! I should've been more active myself - sorry :)
Thanks for merging!

**Pull request merged**